atomic_reactor/dirs.py

```diff
@@ -0,0 +1,227 @@
+        this RootBuildDir object. The caller should not depend on or assume the
+        order how action is applied.
+
+        The action callable might raise errors, that will cause the
+        ``for_each`` terminates immediately and the raised error is propagated
+        to the caller. As a result, the action will not be applied to the rest
+        of the platforms.
+
+        :param action: a callable object that will be applied on every
+            platform-specific directory. This callable must accept one single
+            argument in BuildDir type, and it can return data in any type.
+        :type action: Callable
+        :return: a mapping from platform to the value returned from the
+            function which is called for that platform.
+        :rtype: dict[Platform, any]
+        """
+        results = {}
+        for platform in self.platforms:
+            build_dir = BuildDir(self.path / platform.value, platform)
+            results[platform] = action(build_dir)
+        return results
+
+    def for_all_copy(self, action: Callable[[BuildDir], Iterable[Path]]) -> List[Path]:
+        """Ensure created files are present in all platform-specific directories.
+
+        :param action: a callable that creates files in a given build directory
+            and returns the files it created. The action accepts one single
+            argument in type BuildDir, and returns an iterable object that
+            yields paths of the created files.
+        :type action: callable
+        :return: the list of absolute paths of the created files in all build
+            directories.
+        :rtype: list[pathlib.Path]
+        """
+        first_platform: Platform = self.platforms[0]
+        build_dir = self.path / first_platform.value
+        created_files = action(BuildDir(build_dir, first_platform))
+
+        # Ensure the created files' path is the absolute path, and make some
+        # checks.
+        the_new_files: List[Path] = []
+        filepath: Path
+        for filepath in created_files:
+            if not filepath.is_absolute():
+                filepath = filepath.resolve()
+            if filepath.is_dir():
+                raise ValueError(f"{filepath} is a directory. A normal file is expected.")
+            if not str(filepath).startswith(str(build_dir)):
+                raise ValueError(
+                    f"File must be created inside the build directory. "
+                    f"File {filepath} is not allowed."
+                )
+            the_new_files.append(filepath)
+
+        copied_files: List[Path] = []
+        for platform in self.platforms[1:]:
+            for src_file in the_new_files:
+                dest_build_dir = Path(
+                    str(src_file.parent)
+                    .replace(first_platform.value, platform.value)
+                )
+                dest_build_dir.mkdir(parents=True, exist_ok=True)
+                dest_file = dest_build_dir / src_file.name
```

Comment:
```suggestion
                dest = self.path / platform.value / src_file.relative_to(build_dir)
                dest.parent.mkdir(parents=True, exist_ok=True)
```
Would this be simpler? I mainly wanted to avoid the string replacement, it could cause problems if the src_file path contains platform names in unexpected places